pkg/limits/playback_manager.go

```diff
@@ -0,0 +1,144 @@
+package limits
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/coder/quartz"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	"github.com/grafana/loki/v3/pkg/limits/proto"
+)
+
+// A PlaybackManager processes records from the metadata topic. It is
+// responsible for replaying newly assigned partitions and merging records
+// from other zones.
+type PlaybackManager struct {
+	client           *kgo.Client
+	partitionManager *PartitionManager
+	usage            *UsageStore
+	// readinessCheck checks if a waiting or replaying partition can be
+	// switched to ready.
+	readinessCheck   PartitionReadinessCheck
+	logger           log.Logger
+	recordsProcessed *prometheus.CounterVec
+
+	// Used for tests.
+	clock quartz.Clock
+}
+
+// NewPlaybackManager returns a new PlaybackManager.
+func NewPlaybackManager(
+	client *kgo.Client,
+	partitionManager *PartitionManager,
+	usage *UsageStore,
+	readinessCheck PartitionReadinessCheck,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *PlaybackManager {
+	return &PlaybackManager{
+		client:           client,
+		partitionManager: partitionManager,
+		usage:            usage,
+		readinessCheck:   readinessCheck,
+		logger:           logger,
+		clock:            quartz.NewReal(),
+		recordsProcessed: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "loki_ingest_limits_records_processed_total",
+				Help: "The total number of records processed.",
+			},
+			[]string{"partition"},
+		),
+	}
+}
+
+func (m *PlaybackManager) Run(ctx context.Context) {
+	b := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Second,
+		MaxRetries: 0,
+	})
+	for b.Ongoing() {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if err := m.pollFetches(ctx); err != nil {
+				if errors.Is(err, kgo.ErrClientClosed) {
+					return
+				}
+				level.Error(m.logger).Log("msg", "failed to poll fetches", "err", err.Error())
+				b.Wait()
+			}
+		}
+	}
+}
+
+func (m *PlaybackManager) pollFetches(ctx context.Context) error {
+	fetches := m.client.PollFetches(ctx)
+	if err := fetches.Err(); err != nil {
+		return err
+	}
+	fetches.EachPartition(m.processFetchTopicPartition(ctx))
+	return nil
+}
+
+func (m *PlaybackManager) processFetchTopicPartition(ctx context.Context) func(kgo.FetchTopicPartition) {
```

Comment:
What happens on rebalance with this method? AFAIU the partition lifecycler kicks in and evicts partitions in the partition manager as well as records replayed in the usage store, right?

Is this enough or do we need `BlockRebalanaceOnPoll` to revert the work done in this method and start replay anew after rebalance?

Reply:
There is a race here where a record could be inserted between calling `m.partitionManager.GetState` (which if the partition has been revoked causes all remaining records to be discarded with the message `discarding records for partition as the partition is not assigned to this client`) and `m.usage.Update`. However, I think we can solve this with a.) scheduled evictions of the usage store and b.) evicting the partition also on assignment, ensuring that no records are left if a partition is unassigned and then re-assigned in succession.

`BlockRebalanaceOnPoll` has a number of disadvantages that I think we need to consider carefully before we start to use it.